File: clients/rust/src/protocol/to_client.rs
use serde::{Deserialize, Serialize};
use serde_json::Value as JsonValue;

// Only called for SSE because we don't need this for WebSockets
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Init {
    // Actor ID
    pub ai: String,
    // Connection ID
    pub ci: String,
    // Connection token
    pub ct: String,
}

// Used for connection errors (both during initialization and afterwards)
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Error {
    // Code
    pub c: String,
    // Message
    pub m: String,
    // Metadata
    #[serde(skip_serializing_if = "Option::is_none")]
    pub md: Option<JsonValue>,
    // Action ID
    #[serde(skip_serializing_if = "Option::is_none")]
    pub ai: Option<i64>
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ActionResponse {
    // ID
    pub i: i64,
    // Output
    pub o: JsonValue
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Event {
    // Event name
    pub n: String,
    // Event arguments
    pub a: Vec<JsonValue>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(untagged)]
pub enum ToClientBody {
    Init { i: Init },
    Error { e: Error },
    ActionResponse { ar: ActionResponse },
    EventMessage { ev: Event },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ToClient {
    // Body
    pub b: ToClientBody,
}

File: packages/core/src/actor/protocol/message/to-client.ts
import { z } from "zod";

// Only called for SSE because we don't need this for WebSockets
export const InitSchema = z.object({
	// Actor ID
	ai: z.string(),
	// Connection ID
	ci: z.string(),
	// Connection token
	ct: z.string(),
});

// Used for connection errors (both during initialization and afterwards)
export const ErrorSchema = z.object({
	// Code
	c: z.string(),
	// Message
	m: z.string(),
	// Metadata
	md: z.unknown().optional(),
	// Action ID
	ai: z.number().int().optional(),
});

export const ActionResponseSchema = z.object({
	// ID
	i: z.number().int(),
	// Output
	o: z.unknown(),
});

export const EventSchema = z.object({
	// Name
	n: z.string(),
	// Args
	a: z.array(z.unknown()),
});

export const ToClientSchema = z.object({
	// Body
	b: z.union([
		z.object({ i: InitSchema }),
		z.object({ e: ErrorSchema }),
		z.object({ ar: ActionResponseSchema }),
		z.object({ ev: EventSchema }),
	]),
});

export type ToClient = z.infer<typeof ToClientSchema>;
export type Error = z.infer<typeof ErrorSchema>;
export type ActionResponse = z.infer<typeof ActionResponseSchema>;
export type Event = z.infer<typeof EventSchema>;
